src/main/scala/loamstream/drm/AccountingClient.scala

```diff
@@ -1,13 +1,68 @@
 package loamstream.drm
 
+import scala.concurrent.duration._
+import scala.util.Try
+import scala.util.Success
+import scala.util.Failure
+import loamstream.util.Loggable
+import loamstream.util.Functions
+import loamstream.util.RunResults
+import loamstream.util.Processes
+import loamstream.util.Tries
+import loamstream.util.Loops
+import loamstream.model.execute.Resources.DrmResources
+
 /**
  * @author clint
  * Mar 9, 2017
  *
  * An abstraction for getting some environment-specific metadata that can't currently be accessed via DRMAA
  */
 trait AccountingClient {
-  def getExecutionNode(jobId: String): Option[String]
+  def getResourceUsage(jobId: String): Try[DrmResources]
+}
+
+object AccountingClient extends Loggable {
+  protected[drm] type InvocationFn[A] = A => Try[RunResults] 
+  
+  protected[drm] def doRetries(
+      binaryName: String,
+      maxRetries: Int,
+      delayStart: Duration,
+      delayCap: Duration,
+      delegateFn: InvocationFn[String]): InvocationFn[String] = Functions.memoize { jobId =>
+        
+    val maxRuns = maxRetries + 1
+    
+    def invokeBinary(): Try[RunResults] = delegateFn(jobId) match {
+      case Success(r @ RunResults(executable, exitCode, _, _)) if r.isFailure => {
```

Comment:
While I'm sure this is "right", it feels weird to have a failed `RunResult` end up in the `Success` side of a `Try`? Would it be possible for the `delegateFn` function to return `Failure` for a failed `RunResult` instead?

Reply:
Reading the code a bit more, is the `Try` meant to capture the retries (e.g. `Failure` == too many retries)? Either way, it's a touch confusing. Probably either needs a comment describing how `Try` fits in/is used or something else.

Reply:
I went back and forth on this a lot, and generally agree that the `Try` is confusing.  The thing is that `RunResults` represents the results (stdout/stderr, exit code) of a command that might or might not have succeeded, at least in the exit-code-sense.  `Try`'s notion of success is based on whether or not an exception was thrown somewhere.

That distinction is useful; a `Failure` indicates that something went very wrong that prevented the command from even being run (maybe permissions were wrong, some path didn't exist, etc), as opposed to a command that actually ran but returned a non-zero status code.

I end up coercing the RunResults with non-zero exit codes to `Failure`s anyway, so maybe what's needed is a new ADT to capture the 3 cases: worked-with-a-zero-exit-code, worked-with-a-non-zero-exit-code, and didn't-work-at-all.